dsp/window/window.go

```diff
@@ -0,0 +1,300 @@
+// 0.012604*cos(6*π*k/(N-1)),
+// where N is the length of the window and k=0,1,...,N-1.
+//
+// Spectral leakage parameters: ΔF_0 = 8, ΔF_0.5 = 1.98, K = 4, ɣ_max = -93, β = -9.
+func Nuttall(seq []float64) []float64 {
+	const (
+		a0 = 0.355768
+		a1 = 0.487396
+		a2 = 0.144232
+		a3 = 0.012604
+	)
+
+	k := 2.0 * math.Pi / float64(len(seq)-1)
+	for i := range seq {
+		x := k * float64(i)
+		seq[i] *= a0 - a1*math.Cos(x) + a2*math.Cos(2.0*x) - a3*math.Cos(3.0*x)
+	}
+	return seq
+}
+
+// BlackmanNuttall modifies the seq in place by Blackman-Nuttall window and returns the seq.
+//
+// Blackman-Nuttall window is a low-resolution window
+// (https://www.recordingblogs.com/wiki/blackman-nuttall-window):
+//
+// w[k] = 0.3635819 - 0.4891775*cos(2*π*k/(N-1)) + 0.1365995*cos(4*π*k/(N-1)) -
+// 0.0106411*cos(6*π*k/(N-1)),
+// where N is the length of the window and k=0,1,...,N-1.
+//
+// Spectral leakage parameters: ΔF_0 = 8, ΔF_0.5 = 1.94, K = 4, ɣ_max = -98, β = -8.8.
+func BlackmanNuttall(seq []float64) []float64 {
+	const (
+		a0 = 0.3635819
+		a1 = 0.4891775
+		a2 = 0.1365995
+		a3 = 0.0106411
+	)
+
+	k := 2.0 * math.Pi / float64(len(seq)-1)
+	for i := range seq {
+		x := k * float64(i)
+		seq[i] *= a0 - a1*math.Cos(x) + a2*math.Cos(2.0*x) - a3*math.Cos(3.0*x)
+	}
+	return seq
+}
+
+// FlatTop modifies the seq in place by Flat Top window and returns the seq.
+//
+// Flat Top window is a low-resolution window
+// (https://www.recordingblogs.com/wiki/flat-top-window):
+//
+// w[k] = 0.21557895 - 0.41663158*cos(2*π*k/(N-1)) +
+// 0.277263158*cos(4*π*k/(N-1)) - 0.083578947*cos(6*π*k/(N-1)) +
+// 0.006947368*cos(4*π*k/(N-1)),
+// where N is the length of the window and k=0,1,...,N-1.
+//
+// Spectral leakage parameters: ΔF_0 = 10, ΔF_0.5 = 3.72, K = 5, ɣ_max = -93.0, β = -13.34.
+func FlatTop(seq []float64) []float64 {
+	const (
+		a0 = 0.21557895
+		a1 = 0.41663158
+		a2 = 0.277263158
+		a3 = 0.083578947
+		a4 = 0.006947368
+	)
+
+	k := 2.0 * math.Pi / float64(len(seq)-1)
+	for i := range seq {
+		x := k * float64(i)
+		seq[i] *= a0 - a1*math.Cos(x) + a2*math.Cos(2.0*x) - a3*math.Cos(3.0*x) + a4*math.Cos(4.0*x)
+	}
+	return seq
+}
+
+// Gauss modifies the seq in place by Gauss window and returns the seq.
+//
+// Gauss window is a adjustable window
+// (https://www.recordingblogs.com/wiki/gaussian-window):
+//
+// w[k] = exp(-0.5 * ((k-M)/(σ*M))² ), M = (N-1)/2,
+// where N is the length of the window and k=0,1,...,N-1.
+// The properties of window depends on σ (sigma) argument.
+// It can be used as high or low resolution window, depends of a σ value.
+//
+// Spectral leakage parameters are summarized in the table:
+//         |  σ=0.3 |  σ=0.5 |  σ=1.2 |
+//  -------|--------------------------|
+//  ΔF_0   |  8     |  3.4   |  2.2   |
+//  ΔF_0.5 |  1.82  |  1.2   |  0.94  |
+//  K      |  4     |  1.7   |  1.1   |
+//  ɣ_max  | -65    | -31.5  | -15.5  |
+//  β      | -8.52  | -4.48  | -0.96  |
+func Gauss(seq []float64, sigma float64) []float64 {
```

Comment:
```
// Gaussian modifies seq in place by the Gaussian window and returns the result.
// See https://en.wikipedia.org/wiki/Window_function#Gaussian_window
// and https://www.recordingblogs.com/wiki/gaussian-window for details.
//
// The Gaussian window is an adjustable window.
//
// The sequence weights are
//  w[k] = exp(-0.5 * ((k-M)/(σ*M))² ), M = (N-1)/2,
// for k=0,1,...,N-1 where N is the length of the window.
//
// The properties of window depends on the σ (sigma) argument.
// It can be used as high or low resolution window, depending of the σ value.
//
// Spectral leakage parameters are summarized in the table:
//         |  σ=0.3  |  σ=0.5 |  σ=1.2 |
//  -------|-------------------------=-|
//  ΔF_0   |   8     |   3.4  |   2.2  |
//  ΔF_0.5 |   1.82  |   1.2  |   0.94 |
//  K      |   4     |   1.7  |   1.1  |
//  ɣ_max  | -65     | -31.5  | -15.5  |
//  β      |  -8.52  |  -4.48 |  -0.96 |
func Gaussian(seq []float64, sigma float64) []float64 {
```